File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/command/base/builder/HeaderEnabledCommandBuilder.java
package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.base.builder;

import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.base.Command;

public abstract class HeaderEnabledCommandBuilder<T extends HeaderEnabledCommandBuilder<T, R>, R extends Command> implements CommandBuilder<R> {
    protected Integer uniqueId;
    protected Short sequenceNumber;
    protected boolean multiCommandFlag = false;

    public R build() {
        if (uniqueId == null) {
            throw new IllegalArgumentException("uniqueId can not be null");
        }
        if (sequenceNumber == null) {
            throw new IllegalArgumentException("sequenceNumber can not be null");
        }
        return buildCommand();
    }

    public final T setUniqueId(int uniqueId) {
        this.uniqueId = uniqueId;
        return (T) this;
    }

    public final T setSequenceNumber(short sequenceNumber) {
        this.sequenceNumber = sequenceNumber;
        return (T) this;
    }

    public final T setMultiCommandFlag(boolean multiCommandFlag) {
        this.multiCommandFlag = multiCommandFlag;
        return (T) this;
    }

    protected abstract R buildCommand();
}

File: pump/omnipod-dash/src/main/kotlin/app/aaps/pump/omnipod/dash/driver/pod/command/base/builder/HeaderEnabledCommandBuilder.kt
package app.aaps.pump.omnipod.dash.driver.pod.command.base.builder

import app.aaps.pump.omnipod.dash.driver.pod.command.base.Command

@Suppress("UNCHECKED_CAST")
abstract class HeaderEnabledCommandBuilder<T : HeaderEnabledCommandBuilder<T, R>, R : Command> : CommandBuilder<R> {

    protected var uniqueId: Int? = null
    protected var sequenceNumber: Short? = null
    protected var multiCommandFlag = false

    override fun build(): R {
        requireNotNull(uniqueId) { "uniqueId can not be null" }
        requireNotNull(sequenceNumber) { "sequenceNumber can not be null" }
        return buildCommand()
    }

    fun setUniqueId(uniqueId: Int): T {
        this.uniqueId = uniqueId
        return this as T
    }

    fun setSequenceNumber(sequenceNumber: Short): T {
        this.sequenceNumber = sequenceNumber
        return this as T
    }

    fun setMultiCommandFlag(multiCommandFlag: Boolean): T {
        this.multiCommandFlag = multiCommandFlag
        return this as T
    }

    protected abstract fun buildCommand(): R
}
